From ba563441540faf3a851e6a9bc8b46528a6a9028f Mon Sep 17 00:00:00 2001
From: Lyon Wang <lyon.wang@nxp.com>
Date: Mon, 3 Sep 2018 10:53:34 +0800
Subject: [PATCH 47/91] alsasink: enable eac3 pass-through mode

- For eac3 need using 192K sample rate
- and the sample count for alsa drvier need counted by 2byte per channel/sample

Upstream-Status: Pending
Signed-off-by: Lyon Wang <lyon.wang@nxp.com>
(cherry picked from commit 535a7bbcfc68fc5377b36ea67b6bef81aecbe1e0)
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 ext/alsa/gstalsasink.c | 11 +++++++++--
 ext/alsa/gstalsasink.h |  1 +
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/ext/alsa/gstalsasink.c b/ext/alsa/gstalsasink.c
index ad0a6f9..54cb51f 100644
--- a/ext/alsa/gstalsasink.c
+++ b/ext/alsa/gstalsasink.c
@@ -489,6 +489,8 @@ set_hwparams (GstAlsaSink * alsa)
       no_channels);
   /* set the stream rate */
   rrate = alsa->rate;
+  if ((TRUE== alsa->passthrough) && alsa->iec958 && (alsa->type == GST_AUDIO_RING_BUFFER_FORMAT_TYPE_EAC3))
+    rrate = 192000;
   CHECK (snd_pcm_hw_params_set_rate_near (alsa->handle, params, &rrate, NULL),
       no_rate);
 #ifndef GST_DISABLE_GST_DEBUG
@@ -903,6 +905,7 @@ alsasink_parse_spec (GstAlsaSink * alsa, GstAudioRingBufferSpec * spec)
   alsa->buffer_time = spec->buffer_time;
   alsa->period_time = spec->latency_time;
   alsa->access = SND_PCM_ACCESS_RW_INTERLEAVED;
+  alsa->type = spec->type;
 
   if ((spec->type == GST_AUDIO_RING_BUFFER_FORMAT_TYPE_RAW ||
           spec->type == GST_AUDIO_RING_BUFFER_FORMAT_TYPE_DSD) &&
@@ -1115,8 +1118,12 @@ gst_alsasink_write (GstAudioSink * asink, gpointer data, guint length)
 
   GST_LOG_OBJECT (asink, "received audio samples buffer of %u bytes", length);
 
-  cptr = length / alsa->bpf;
-
+  if (alsa->iec958 && (TRUE == alsa->passthrough) ) {
+    cptr = length / 4;  // treated as 2ch with S16 in pass-through mode 
+  }
+  else {
+    cptr = length / alsa->bpf;
+  }
   GST_ALSA_SINK_LOCK (asink);
   while (cptr > 0) {
     /* start by doing a blocking wait for free space. Set the timeout
diff --git a/ext/alsa/gstalsasink.h b/ext/alsa/gstalsasink.h
index ab642ee..25567f6 100644
--- a/ext/alsa/gstalsasink.h
+++ b/ext/alsa/gstalsasink.h
@@ -83,6 +83,7 @@ struct _GstAlsaSink {
   GMutex delay_lock;
 
   gboolean passthrough;
+  GstAudioRingBufferFormatType  type;
 };
 
 struct _GstAlsaSinkClass {
-- 
2.43.0

